### PR TITLE
Use HeaderBlacklist to filter unwated headers in reverse proxy

### PR DESCRIPTION
**Pull request by akashshinde**
This PR provides a way to delete unwated headers from proxy response.

`Proxy.Config` already has a field(`HeaderBlacklist`) to blacklist/filter headers, but It is not currently used anywhere. ref: https://github.com/openshift/console/blob/ca7021b644f481f82c0c10aac1df7fa12d8117bd/pkg/proxy/proxy.go#L23
This PR deletes headers provided in the `HeaderBlacklist` 

- [x] Added Unit test

### CODE REVIEW

File: pkg/proxy/proxy.go [hunk 1]
```diff
@@ -54,7 +55,9 @@ func NewProxy(cfg *Config) *Proxy {
 	reverseProxy := httputil.NewSingleHostReverseProxy(cfg.Endpoint)
 	reverseProxy.FlushInterval = time.Millisecond * 100
 	reverseProxy.Transport = transport
-	reverseProxy.ModifyResponse = filterHeaders
+	reverseProxy.ModifyResponse = func(r *http.Response) error {
```
Comment (spadgett) [inline pkg/proxy/proxy.go:58]: This is intended to remove request headers, not response headers.

**Comment by spadgett** [comment]
It looks like we have the request headers hard-coded here, so we're not sending anything we shouldn't.

https://github.com/spadgett/console/blob/bc9a54150429b081278e2024fb6cb769f5ca3a86/pkg/proxy/proxy.go#L88

Maybe we can remove the `HeaderBlacklist` in the proxy config?

**Comment by spadgett** [comment]
@akashshinde Do you have a need to filter additional response headers or did you just notice it was unused?

**Comment by akashshinde** [comment]
@spadgett 
We have a use case where proxied url copies all headers along with response which is correct but origin url/proxied url copies `Cache-Control` header which caches the response in the browser.
Therefore when UI makes second request to same proxy url, browser uses the cached response instead of the one sent by server.
So we want a way to filter out that header.
If we can't use `HeaderBlacklist` then we need to find alternative way to tackle this issue.
Let me know what do you suggest.

**Comment by akashshinde** [comment]
> It looks like we have the request headers hard-coded here, so we're not sending anything we shouldn't.
> 
> https://github.com/spadgett/console/blob/bc9a54150429b081278e2024fb6cb769f5ca3a86/pkg/proxy/proxy.go#L88
> 
> Maybe we can remove the `HeaderBlacklist` in the proxy config?

Please check https://github.com/spadgett/console/blob/bc9a54150429b081278e2024fb6cb769f5ca3a86/cmd/bridge/main.go#L328
We pass `Cookie` and `X-CSRFToken` here in `HeaderBlacklist` (same headers we've hard coded in https://github.com/spadgett/console/blob/bc9a54150429b081278e2024fb6cb769f5ca3a86/pkg/proxy/proxy.go#L88)

So I'm guessing we should remove hard coded headers and keep the field `HeaderBlacklist`, This way we will have control over response headers.

**Comment by spadgett** [comment]
> Please check https://github.com/spadgett/console/blob/bc9a54150429b081278e2024fb6cb769f5ca3a86/cmd/bridge/main.go#L328
> We pass `Cookie` and `X-CSRFToken` here in `HeaderBlacklist` (same headers we've hard coded in https://github.com/spadgett/console/blob/bc9a54150429b081278e2024fb6cb769f5ca3a86/pkg/proxy/proxy.go#L88)

Right these are definitely request headers, not response headers. `X-CSRFToken` comes from the console client, and `Cookie` is a request header.

> So I'm guessing we should remove hard coded headers and keep the field `HeaderBlacklist`, This way we will have control over response headers.

But this was intended to filter request headers. I would rather have the hard-coded list to be honest since we should *never* send these for security reasons. It would be too easy to forget to set that on one of our proxies.

Do you have a specific need to filter a different response header?

**Comment by akashshinde** [comment]
> Do you have a specific need to filter a different response header?

this is the reason - https://github.com/openshift/console/pull/6044#issuecomment-661885925 
Let me know if I am not clear.

**Comment by spadgett** [comment]
I'd make this `ReponseHeaderDenylist` and remove `HeaderBlacklist` from the config and the references in main.go.

**Comment by spadgett** [comment]
The problem though is that removing `Cache-Control` is not enough. It leaves it up to the browser. You really want to replace it with another value.

**Comment by spadgett** [comment]
Any chance of fixing the Cache-Control header in the upstream service if it shouldn't be cached?

**Comment by akashshinde** [comment]
> Any chance of fixing the Cache-Control header in the upstream service if it shouldn't be cached?

https://redhat-developer.github.io/redhat-helm-charts/index.yaml this is the URL we are serving through the proxy and it's being served through GitHub pages. I doubt If we could remove/configure http headers in GitHub service @sbose78 wdyt?
Github page adds two headers which are causing this issue.
```
Cache-Control: 600
Etag: unique-id
```

> The problem though is that removing Cache-Control is not enough. It leaves it up to the browser. You really want to replace it with another value.

Well I was thinking of replacing `Cache-Control` default value to `no-cache, private`, this would avoid caching in browser, But to modify the response headers we would need to add `ModifyResponse func(r *http.Response)` instead of `HeaderBlacklist string[]` in `Config`

`Config` struct would look like 

```
type Config struct {
	ModifyHTTPProxyResponse func(r *http.Response)
	Endpoint        *url.URL
	TLSClientConfig *tls.Config
	Origin          string
}
```

**Comment by spadgett** [comment]
@akashshinde If GitHub pages is setting a valid ETag on the response, why is the Cache-Control header a problem? You should get the cached version only when the ETag hasn't changed.

**Comment by akashshinde** [comment]
@spadgett we are no longer using proxy in https://github.com/openshift/console/pull/5933, therefore we can close this PR unless you have a reason to keep it open

**Comment by sbose78** [comment]
Yea, let's close this PR.

**Pull request closed without merging**